aptos-move/block-executor/src/scheduler_status.rs

```diff
@@ -0,0 +1,1387 @@
+        // invoked once per incarnation, and thus incarnations must always match.
+        if inner_status.incarnation() != finished_incarnation {
+            return Err(code_invariant_error(format!(
+                "Finish execution of incarnation {}, but inner status {:?}",
+                finished_incarnation, inner_status,
+            )));
+        }
+
+        match inner_status.status {
+            StatusEnum::Executing => {
+                inner_status.status = StatusEnum::Executed;
+
+                let new_flag = if self.num_stalls.load(Ordering::Relaxed) == 0 {
+                    dependency_flags::SAFE
+                } else {
+                    dependency_flags::DEFER
+                };
+                self.swap_dependency_flag_any(
+                    &[dependency_flags::EXECUTING],
+                    new_flag,
+                    "finish_execution",
+                )?;
+
+                Ok(true)
+            },
+            StatusEnum::Aborted => {
+                self.incarnate(inner_status, finished_incarnation + 1, false);
+                Ok(false)
+            },
+            StatusEnum::PendingScheduling | StatusEnum::Executed => {
+                Err(code_invariant_error(format!(
+                    "Status update to Executed failed, previous inner status {:?}",
+                    inner_status
+                )))
+            },
+        }
+    }
+
+    /// Attempts to mark a transaction incarnation for abortion.
+    ///
+    /// This is the first step of the two-step abort process. It serves as an efficient
+    /// test-and-set filter for abort attempts, ensuring only one worker successfully
+    /// initiates the abort of a specific incarnation.
+    ///
+    /// # Parameters
+    /// - `incarnation`: The incarnation number to abort
+    ///
+    /// # Returns
+    /// - `Ok(true)` if this call successfully marked the incarnation for abortion
+    /// - `Ok(false)` if the incarnation was already marked for abortion
+    /// - `Err` if the provided incarnation is invalid (greater than the current value)
+    pub(crate) fn try_abort(&self, incarnation: Incarnation) -> Result<bool, PanicError> {
+        let prev_value = self
+            .next_incarnation_to_try_abort
+            .fetch_max(incarnation + 1, Ordering::Relaxed);
+        match incarnation.cmp(&prev_value) {
+            cmp::Ordering::Less => Ok(false),
+            cmp::Ordering::Equal => Ok(true),
+            cmp::Ordering::Greater => Err(code_invariant_error(format!(
+                "Try abort incarnation {} > self.next_incarnation_to_try_abort = {}",
+                incarnation, prev_value,
+            ))),
+        }
+    }
+
+    /// Completes the abortion of a transaction incarnation.
+    ///
+    /// This is the second step of the two-step abort process. It must be called
+    /// after a successful try_abort and updates the transaction's status:
```

Comment:
Otherwise it will return `code_invariant_error`?

Reply:
yep, otherwise invariants are broken. Added a line here (and Err return in the comment also explains)